// cytosis/dist/cytosis_prev-3.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _construct2 = _interopRequireDefault(require("@babel/runtime/helpers/construct"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var Airtable = _interopRequireWildcard(require("airtable"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createForOfIteratorHelper(o) { if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (o = _unsupportedIterableToArray(o))) { var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var it, normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var Cytosis = function () {
  function Cytosis(opts) {
    (0, _classCallCheck2.default)(this, Cytosis);

    var _this = this;

    if (arguments.length === 0) {
      _this.routeName = Cytosis.routeName || '';
      _this.config = Cytosis.config || '';
      _this.endpointUrl = Cytosis.endpointUrl || '';
      _this.airKey = Cytosis.airKey || '';
      _this.airBase = Cytosis.airBase || '';
      _this.airBase.tables = Cytosis.tables || [];
      _this.airBase.tableQuery = Cytosis.tableQuery || undefined;
      _this.airBase.options = Cytosis.options || {
        view: "Grid view"
      };
      _this.airBase.payloads = Cytosis.payloads;
      _this.airBase.endpointUrl = Cytosis.endpointUrl;
    } else {
      _this.routeName = opts.routeName;
      _this.config = opts.config;
      _this.endpointUrl = opts.endpointUrl;
      _this.airKey = opts.airKey;
      _this.airBase = {
        id: opts.airBase || opts.airBaseId
      };
      _this.airBase.tables = opts.tables || [];
      _this.airBase.tableQuery = opts.tableQuery || undefined;
      _this.airBase.options = opts.options || {
        view: "Grid view"
      };
      _this.airBase.payloads = opts.payloads;
    }

    Airtable.configure({
      endpointUrl: _this.endpointUrl || "https://api.airtable.com",
      apiKey: _this.airKey
    });
    _this.base = Cytosis.getBase(this.airBase.id);
    if (!_this.airBase.tableQuery) return this;
    return new Promise(function (resolve, reject) {
      _this.init(_this.airBase.tableQuery).then(function (result) {
        if (result) {
          Cytosis.getTables({
            options: opts.options,
            payloads: opts.payloads,
            cytosis: _this,
            tables: _this.airBase.tables,
            routeName: _this.routeName
          }).then(function (_result) {
            _this.tables = _result;
            resolve(_this);
          }, function (err) {
            reject(new Error("[Cytosis] Cytosis initialization error: Couldn't retrieve all tables from Airtable!", err));
          });
        } else {
          if (!result) reject(result);
          resolve(result);
        }
      }, function (err) {
        reject(new Error("[Cytosis] Cytosis initialization error: Couldn't setup Config (_cytosis)!", err));
      });
    });
  }

  (0, _createClass2.default)(Cytosis, [{
    key: "init",
    value: function init(tableQuery) {
      var reset = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

      var _this = this;

      return new Promise(function (resolve, reject) {
        if (!reset && _this.airBase.tables && _this.airBase.tables.length > 0) {
          resolve(_this);
        }

        if (_this.airBase.tables && _this.airBase.tables.length > 0) resolve(_this);

        var setup = function setup(_config) {
          _this['config'] = _config;

          if (!_this.airBase.tables || _this.airBase.tables.length == 0) {
            var _iterator = _createForOfIteratorHelper(_config._cytosis),
                _step;

            try {
              for (_iterator.s(); !(_step = _iterator.n()).done;) {
                var config = _step.value;

                if (config.fields['Name'] == tableQuery && config.fields['Tables']) {
                  var options = {
                    fields: config.fields['fields'],
                    filter: config.fields['filterByFormula'],
                    maxRecords: config.fields['maxRecords'],
                    pageSize: config.fields['pageSize'],
                    sort: config.fields['sort'] ? JSON.parse(config.fields['sort'])['sort'] : undefined,
                    view: config.fields['view'],
                    matchKeywordWithField: config.fields['matchKeywordWithField'],
                    matchStyle: config.fields['matchStyle']
                  };
                  _this.airBase['tables'] = config.fields['Tables'];
                  _this.airBase['options'] = options;
                } else if (config.fields['Name'] == tableQuery && config.fields['LinkedQueryNames']) {
                    (function () {
                      var linkedQueries = config.fields['LinkedQueryNames'];
                      var tables = [];
                      linkedQueries.map(function (linkedquery) {
                        _config._cytosis.map(function (query) {
                          if (linkedquery == query.fields['Name']) {
                            var _options = {
                              fields: query.fields['fields'],
                              filter: query.fields['filterByFormula'],
                              maxRecords: query.fields['maxRecords'],
                              pageSize: query.fields['pageSize'],
                              sort: query.fields['sort'] ? JSON.parse(query.fields['sort'])['sort'] : undefined,
                              view: query.fields['view']
                            };
                            tables.push({
                              query: linkedquery,
                              tables: query.fields['Tables'],
                              options: _options
                            });
                          }
                        });
                      });
                      _this.airBase['tables'] = tables;
                    })();
                  }
              }
            } catch (err) {
              _iterator.e(err);
            } finally {
              _iterator.f();
            }
          }
        };

        if (_this.config) {
          setup(_this.config);
          resolve(_this);
        } else {
          Cytosis.getTables({
            options: {},
            cytosis: _this,
            tables: ['_cytosis'],
            routeName: _this.routeName
          }).then(function (_config) {
            if (_config) {
              setup(_config);
            }

            if (!_config || !_this.airBase.tables || _this.airBase.tables.length == 0) {
              reject(new Error("[Cytosis] \u2014 couldn\u2019t find a '_cytosis' table with row [query:".concat(tableQuery, "] or 'tables' filled out with the names of tables to load")));
            }

            resolve(_this);
          }, function (err) {
            reject(new Error("[Cytosis] Couldn't retrieve Config object from Airtable", err));
          });
        }
      });
    }
  }, {
    key: "find",
    value: function find(findStr) {
      var fields = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : ['Name'];
      return Cytosis.find(findStr, this.tables, fields);
    }
  }, {
    key: "save",
    value: function save(object, tableName) {
      var recordId = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : undefined;
      return Cytosis.save(object, tableName, this, recordId);
    }
  }, {
    key: "saveArray",
    value: function saveArray(objectArray, tableName) {
      var create = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      var typecast = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
      return Cytosis.saveArray(objectArray, tableName, this, create, typecast);
    }
  }, {
    key: "delete",
    value: function _delete(tableName, recordId) {
      return Cytosis.delete(tableName, this, recordId);
    }
  }, {
    key: "saveLinkedTable",
    value: function saveLinkedTable(stringList, targetTableName, sourceTable) {
      var colName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 'Name';
      return Cytosis.saveLinkedTable(stringList, targetTableName, sourceTable, this, colName);
    }
  }], [{
    key: "getBase",
    value: function getBase(baseId) {
      return Airtable.base(baseId);
    }
  }, {
    key: "preCheck",
    value: function preCheck(airKey, airBase) {
      if (airKey && airBase) return true;
      return false;
    }
  }, {
    key: "getTables",
    value: function getTables(_ref) {
      var options = _ref.options,
          payloads = _ref.payloads,
          cytosis = _ref.cytosis,
          tables = _ref.tables,
          routeName = _ref.routeName;
      options = options || cytosis.airBase.options || {};
      tables = tables || cytosis.airBase.tables;
      var pTables = [];
      if (!Cytosis.preCheck(cytosis.airKey, cytosis.airBase)) return {};

      function getTablePromise(_ref2) {
        var tables = _ref2.tables,
            options = _ref2.options,
            payloads = _ref2.payloads;

        try {
          var view = options.view,
              fields = options.fields,
              sort = options.sort,
              filter = options.filter,
              maxRecords = options.maxRecords,
              pageSize = options.pageSize;
          if (!view) view = '';
          if (!filter) filter = '';

          if (payloads && payloads.keyword && options && options.matchKeywordWithField) {
            filter = "IF({".concat(options.matchKeywordWithField, "} = \"").concat(payloads.keyword, "\",TRUE(),FALSE())");
            if (options.matchStyle == "partial") filter = "IF(SEARCH(\"".concat(payloads.keyword, "\",{").concat(options.matchKeywordWithField, "}) > 0,TRUE(),FALSE())");
          }

          var _iterator2 = _createForOfIteratorHelper(tables),
              _step2;

          try {
            var _loop = function _loop() {
              var table = _step2.value;
              var list = [];
              var filterObj = {
                filterByFormula: filter,
                view: view
              };

              if (sort) {
                filterObj['sort'] = sort;
              }

              if (maxRecords) {
                filterObj['maxRecords'] = maxRecords;
              }

              if (pageSize) {
                filterObj['pageSize'] = pageSize;
              }

              if (fields && fields[table]) {
                filterObj['fields'] = fields[table];
              } else if (fields) {
                filterObj['fields'] = fields;
              }

              var airtableFetch = function airtableFetch() {
                return new Promise(function (resolve, reject) {
                  var base = Cytosis.getBase(cytosis.airBase.id);
                  var pageCount = 1;
                  base(table).select(filterObj).eachPage(function page(records, fetchNextPage) {
                    console.log('[Cytosis] Page Fetch for:', table, 'routeName:', routeName, 'page:', pageCount);
                    pageCount += 1;
                    records.forEach(function (record) {
                      list.push(record);
                    });
                    var delay = 250;
                    setTimeout(fetchNextPage, delay);
                  }, function done(err) {
                    if (err) {
                      console.error('[Cytosis/getTablePromise/airtableFetch] Airtable Fetch Error @routeName:', routeName);
                      console.error('[Cytosis/getTablePromise/airtableFetch] Airtable Fetch Error [2]', 'Errored on table:', table, 'tablesLen:', tables.length, 'tables:', tables);
                      console.error('[Cytosis/getTablePromise/airtableFetch] Airtable Fetch Error >> error message:', err);
                      reject(new Error("[Cytosis/getTablePromise/airtableFetch] No response from Airtable"));
                    }

                    resolve((0, _defineProperty2.default)({}, table, list));
                  });
                });
              };

              pTables.push(airtableFetch());
            };

            for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
              _loop();
            }
          } catch (err) {
            _iterator2.e(err);
          } finally {
            _iterator2.f();
          }
        } catch (e) {
          console.error('[Cytosis/getTables/getTablePromise] Airtable caught general error', e);
        }
      }

      if (typeof tables[0] == 'string') {
        getTablePromise({
          tables: tables,
          options: options,
          payloads: payloads
        });
      } else {
          tables.map(function (query) {
            getTablePromise({
              tables: query.tables,
              options: query.options,
              payloads: payloads
            });
          });
        }

      try {
        return Promise.all(pTables).then(function (tables) {
          var finalObj = {};

          var _iterator3 = _createForOfIteratorHelper(tables),
              _step3;

          try {
            for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
              var t = _step3.value;
              finalObj = _objectSpread(_objectSpread(_objectSpread({}, finalObj), t), cytosis.data);
            }
          } catch (err) {
            _iterator3.e(err);
          } finally {
            _iterator3.f();
          }

          return finalObj;
        }, function (err) {
          console.error("[Cytosis/getTables] A table errored out or timed out: ", err);
          return Promise.reject(new Error("[Cytosis/getTables] Fetch Error"));
        });
      } catch (err) {
        console.error("[Cytosis/getTables/pTablesPromiseHandling] An Airtable table errored out", err);
      }
    }
  }, {
    key: "getConfig",
    value: function getConfig(_ref3) {
      var airKey = _ref3.airKey,
          airBase = _ref3.airBase,
          routeName = _ref3.routeName,
          endpointUrl = _ref3.endpointUrl;
      Airtable.configure({
        endpointUrl: endpointUrl || "https://api.airtable.com",
        apiKey: airKey
      });
      var configP = Cytosis.getTables({
        cytosis: {
          'airKey': airKey,
          'airBase': {
            'id': airBase
          }
        },
        tables: ['_cytosis'],
        routeName: '[_getConfig] ' + routeName
      });
      return configP;
    }
  }, {
    key: "get",
    value: function get(recordId, tables) {
      var result;

      if (tables) {
        Object.keys(tables).map(function (table) {
          var _iterator4 = _createForOfIteratorHelper(tables[table]),
              _step4;

          try {
            for (_iterator4.s(); !(_step4 = _iterator4.n()).done;) {
              var record = _step4.value;

              if (record.id == recordId) {
                result = record;
              }
            }
          } catch (err) {
            _iterator4.e(err);
          } finally {
            _iterator4.f();
          }
        });
        return result;
      }

      return undefined;
    }
  }, {
    key: "getRecord",
    value: function () {
      var _getRecord = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(_ref4) {
        var recordId, base, tableName, airKey, baseId, endpointUrl, record;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                recordId = _ref4.recordId, base = _ref4.base, tableName = _ref4.tableName, airKey = _ref4.airKey, baseId = _ref4.baseId, endpointUrl = _ref4.endpointUrl;
                _context.prev = 1;

                if (!base) {
                  Airtable.configure({
                    endpointUrl: endpointUrl || "https://api.airtable.com",
                    apiKey: airKey
                  });
                  base = Cytosis.getBase(baseId);
                }

                _context.next = 5;
                return base(tableName).find(recordId);

              case 5:
                record = _context.sent;
                return _context.abrupt("return", record);

              case 9:
                _context.prev = 9;
                _context.t0 = _context["catch"](1);
                return _context.abrupt("return", Promise.reject());

              case 12:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, null, [[1, 9]]);
      }));

      function getRecord(_x) {
        return _getRecord.apply(this, arguments);
      }

      return getRecord;
    }()
  }, {
    key: "find",
    value: function find(findStr, tables) {
      var fields = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : ['Name'];
      if (!findStr || !tables) return [];

      if (typeof fields == "string") {
        console.error('[Cytosis] find "fields" argument must be an array');
        return undefined;
      }

      function matchField(str, tables, fields) {
        var results = [];
        Object.keys(tables).map(function (table) {
          if (!tables[table]) throw new Error("[Cytosis/Find] \u2014 Couldn\u2019t find a match. Make sure you're looking in the right place. Reference table/string: (".concat(tables[table], " / ").concat(findStr, "). Required Format was probably wrong: { Content: [ row, row, row], Tags: [ row, row, row ] }. "));

          var _iterator5 = _createForOfIteratorHelper(tables[table]),
              _step5;

          try {
            for (_iterator5.s(); !(_step5 = _iterator5.n()).done;) {
              var record = _step5.value;

              var _iterator6 = _createForOfIteratorHelper(fields),
                  _step6;

              try {
                for (_iterator6.s(); !(_step6 = _iterator6.n()).done;) {
                  var field = _step6.value;

                  if (record && record.fields && record.fields[field] && str == record.fields[field]) {
                    results.push(record);
                  }
                }
              } catch (err) {
                _iterator6.e(err);
              } finally {
                _iterator6.f();
              }
            }
          } catch (err) {
            _iterator5.e(err);
          } finally {
            _iterator5.f();
          }
        });
        return results;
      }

      var queries = findStr.split('.');
      if (queries.length == 1) return matchField(queries[0], tables, fields)[0];
      if (queries.length == 2) return matchField(queries[1], {
        q: tables[queries[0]]
      }, fields);
      var records = matchField(queries[1], {
        q: tables[queries[0]]
      }, fields);

      if (!Array.isArray(records[0].fields[queries[2]])) {
        return records[0].fields[queries[2]];
      }

      var fieldContent = records[0].fields[queries[2]];

      if (queries.length == 4) {
        var result = [];

        var _iterator7 = _createForOfIteratorHelper(fieldContent),
            _step7;

        try {
          for (_iterator7.s(); !(_step7 = _iterator7.n()).done;) {
            var id = _step7.value;
            var record = Cytosis.get(id, tables);
            result.push(record.fields[queries[3]]);
          }
        } catch (err) {
          _iterator7.e(err);
        } finally {
          _iterator7.f();
        }

        return result.join(', ');
      }

      return fieldContent;
    }
  }, {
    key: "findOne",
    value: function findOne(findStr, table) {
      var fields = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : ['Name'];
      if (!table) return undefined;

      if (typeof fields == "string") {
        console.error('[Cytosis] find "fields" argument must be an array');
        return undefined;
      }

      var key = findStr.split('.').length > 0 ? findStr.split('.')[0] : '_key';
      var payload = {};
      payload[key] = table;
      var output = this.find(findStr, payload, fields);
      if (output && output.length && output.length > 0) return output[0];else if (output) return output;
      return undefined;
    }
  }, {
    key: "findField",
    value: function findField(findStr, table, contentField) {
      var fields = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : ['Name'];
      var element = Cytosis.findOne(findStr, table, fields);
      if (element && element.fields && element.fields[contentField]) return element.fields[contentField];
      return undefined;
    }
  }, {
    key: "save",
    value: function save(object, tableName, cytosis) {
      var recordId = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : undefined;
      if (!Cytosis.preCheck(cytosis.airKey, cytosis.airBase)) return;
      var base = cytosis.base;

      try {
        return new Promise(function (resolve, reject) {
          if (!recordId) {
            base(tableName).create(object, function (err, record) {
              if (err) {
                console.error('Airtable async save/create error', err);
                reject(err);
                return;
              }

              console.log('New record: ', record.getId(), record.fields['Name']);
              resolve(record);
            });
          } else {
            base(tableName).update(recordId, object, function (err, record) {
              if (err) {
                console.error('Airtable async save error', err);
                reject(err);
                return;
              }

              console.log('Updated record: ', record.getId(), record.fields['Name']);
              resolve(record);
            });
          }
        });
      } catch (e) {
        console.error('Save Object to Airtable error (do you have permission?)', e);
        return;
      }
    }
  }, {
    key: "saveArray",
    value: function saveArray(objectArray, tableName, cytosis) {
      var create = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
      var typecast = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : false;
      if (!Cytosis.preCheck(cytosis.airKey, cytosis.airBase)) return;
      var base = cytosis.base;

      try {
        return new Promise(function (resolve, reject) {
          if (create) {
            base(tableName).create(objectArray, {
              typecast: typecast
            }, function (err, records) {
              if (err) {
                console.error('Airtable async saveArray/create error', err);
                reject(err);
                return;
              }

              console.log('New records: ', records);
              resolve(records);
            });
          } else {
            base(tableName).update(objectArray, {
              typecast: typecast
            }, function (err, records) {
              if (err) {
                console.error('Airtable async saveArray/update error', err);
                reject(err);
                return;
              }

              console.log('Updated records: ', records);
              resolve(records);
            });
          }
        });
      } catch (e) {
        console.error('SaveArray Object to Airtable error (do you have permission?)', e);
        return;
      }
    }
  }, {
    key: "delete",
    value: function _delete(tableName, cytosis, recordId) {
      if (!Cytosis.preCheck(cytosis.airKey, cytosis.airBase)) return;
      var base = cytosis.base;

      try {
        return new Promise(function (resolve, reject) {
          if (recordId) {
            base(tableName).destroy(recordId, function (err, record) {
              if (err) {
                console.error('Airtable async delete error', err);
                reject(err);
                return;
              }

              console.log('Deleted record: ', record.getId(), record.fields['Name']);
              resolve(record);
            });
          }
        });
      } catch (e) {
        console.error('Delete Object from Airtable error (do you have permission?)', e);
        return;
      }
    }
  }, {
    key: "saveLinkedTable",
    value: function () {
      var _saveLinkedTable = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee3(stringList, targetTableName, sourceTable, cytosis) {
        var colName,
            recordIds,
            _args3 = arguments;
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                colName = _args3.length > 4 && _args3[4] !== undefined ? _args3[4] : 'Name';
                _context3.next = 3;
                return stringList.reduce(function () {
                  var _ref5 = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee2(resultPromise, listItem) {
                    var _result, _iterator8, _step8, record, recordName, recordId;

                    return _regenerator.default.wrap(function _callee2$(_context2) {
                      while (1) {
                        switch (_context2.prev = _context2.next) {
                          case 0:
                            _context2.next = 2;
                            return resultPromise;

                          case 2:
                            _result = _context2.sent;
                            _iterator8 = _createForOfIteratorHelper(sourceTable);
                            _context2.prev = 4;

                            _iterator8.s();

                          case 6:
                            if ((_step8 = _iterator8.n()).done) {
                              _context2.next = 13;
                              break;
                            }

                            record = _step8.value;
                            recordName = record.fields[colName];

                            if (!(recordName && listItem.toLowerCase() == recordName.toLowerCase())) {
                              _context2.next = 11;
                              break;
                            }

                            return _context2.abrupt("return", _result.concat(record.getId()));

                          case 11:
                            _context2.next = 6;
                            break;

                          case 13:
                            _context2.next = 18;
                            break;

                          case 15:
                            _context2.prev = 15;
                            _context2.t0 = _context2["catch"](4);

                            _iterator8.e(_context2.t0);

                          case 18:
                            _context2.prev = 18;

                            _iterator8.f();

                            return _context2.finish(18);

                          case 21:
                            _context2.next = 23;
                            return Cytosis.save({
                              'Name': listItem
                            }, targetTableName, cytosis);

                          case 23:
                            recordId = _context2.sent;
                            return _context2.abrupt("return", _result.concat(recordId.id));

                          case 25:
                          case "end":
                            return _context2.stop();
                        }
                      }
                    }, _callee2, null, [[4, 15, 18, 21]]);
                  }));

                  return function (_x6, _x7) {
                    return _ref5.apply(this, arguments);
                  };
                }(), Promise.resolve([]));

              case 3:
                recordIds = _context3.sent;
                return _context3.abrupt("return", recordIds);

              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3);
      }));

      function saveLinkedTable(_x2, _x3, _x4, _x5) {
        return _saveLinkedTable.apply(this, arguments);
      }

      return saveLinkedTable;
    }()
  }, {
    key: "cleanTable",
    value: function cleanTable(table) {
      return table.map(function (entry) {
        return {
          fields: entry.fields,
          id: entry.id
        };
      });
    }
  }, {
    key: "cleanRecord",
    value: function cleanRecord(record) {
      return {
        fields: record.fields,
        id: record.id
      };
    }
  }, {
    key: "strip",
    value: function strip(cytosis) {
      var _cytosis = {};
      _cytosis['config'] = {
        _cytosis: Cytosis.cleanTable(cytosis.config._cytosis)
      };
      _cytosis['airBase'] = cytosis['airBase'];
      _cytosis['airKey'] = cytosis['airKey'];
      _cytosis['endpointUrl'] = cytosis['endpointUrl'];
      _cytosis['routeName'] = cytosis['routeName'];
      _cytosis['tables'] = {};
      Object.keys(cytosis.tables).map(function (tableName) {
        _cytosis['tables'][tableName] = Cytosis.cleanTable(cytosis.tables[tableName]);
      });
      return _cytosis;
    }
  }, {
    key: "blankFields",
    value: function blankFields() {
      if ((0, _typeof2.default)(this.fields) !== 'object' || this.fields === null) return {};
      var entries = Object.entries(this.fields);
      var blankFields = {};

      for (var _i = 0, _entries = entries; _i < _entries.length; _i++) {
        var _entries$_i = (0, _slicedToArray2.default)(_entries[_i], 2),
            key = _entries$_i[0],
            settings = _entries$_i[1];

        console.log('entries:', entries, key, settings);

        if (settings === undefined) {
          var error = new Error("Improper Field Definition in Table '".concat(this.name, "'.\n") + "Received: ".concat(settings));
          error.name = 'TableError';
          throw error;
        }

        if (typeof settings.name !== 'string') {
          var _error = new Error("Improper Field Definition in Table '".concat(this.name, "'.\n") + "Expected 'name' to be a string.\n" + "Received: ".concat(settings));

          _error.name = 'TableError';
          throw _error;
        }

        var args = [settings.name, undefined, _objectSpread({}, settings)];
        var blankField = typeof settings.type !== 'function' ? (0, _construct2.default)(UnknownField, args) : (0, _construct2.default)(settings.type, args);
        blankFields[key] = blankField;
      }

      return blankFields;
    }
  }, {
    key: "getLinkedRecords",
    value: function getLinkedRecords(recordIdArray, sourceArray) {
      var getObj = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      if (!recordIdArray || !sourceArray) return [];
      var records = [];

      var _iterator9 = _createForOfIteratorHelper(recordIdArray),
          _step9;

      try {
        for (_iterator9.s(); !(_step9 = _iterator9.n()).done;) {
          var recordId = _step9.value;

          var _iterator10 = _createForOfIteratorHelper(sourceArray),
              _step10;

          try {
            for (_iterator10.s(); !(_step10 = _iterator10.n()).done;) {
              var linkedRecord = _step10.value;

              if (recordId == linkedRecord.id) {
                if (getObj) {
                  records.push(linkedRecord);
                } else {
                  records.push(linkedRecord.fields['Name']);
                }
              }
            }
          } catch (err) {
            _iterator10.e(err);
          } finally {
            _iterator10.f();
          }
        }
      } catch (err) {
        _iterator9.e(err);
      } finally {
        _iterator9.f();
      }

      return records;
    }
  }, {
    key: "getFieldContent",
    value: function getFieldContent(recordArray, fieldName) {
      var linkedTable = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : undefined;
      var results = [];

      var _iterator11 = _createForOfIteratorHelper(recordArray),
          _step11;

      try {
        for (_iterator11.s(); !(_step11 = _iterator11.n()).done;) {
          var record = _step11.value;

          if (linkedTable) {
            var recordIds = record.fields[fieldName];
            var linked = Cytosis.getLinkedRecords(recordIds, linkedTable, true);
            if (linked.length > 0) results = results.concat(linked);else results.push(record.fields[fieldName]);
          } else results.push(record.fields[fieldName]);
        }
      } catch (err) {
        _iterator11.e(err);
      } finally {
        _iterator11.f();
      }

      return results;
    }
  }, {
    key: "getFieldValues",
    value: function getFieldValues(recordArray, field) {
      var results = [];

      var _iterator12 = _createForOfIteratorHelper(recordArray),
          _step12;

      try {
        var _loop2 = function _loop2() {
          var record = _step12.value;
          var recordValue = record.fields[field];

          if (Array.isArray(recordValue)) {
            var _iterator13 = _createForOfIteratorHelper(recordValue),
                _step13;

            try {
              var _loop3 = function _loop3() {
                var rV = _step13.value;

                if (!results.find(function (r) {
                  return r == rV;
                })) {
                  results.push(rV);
                }
              };

              for (_iterator13.s(); !(_step13 = _iterator13.n()).done;) {
                _loop3();
              }
            } catch (err) {
              _iterator13.e(err);
            } finally {
              _iterator13.f();
            }
          } else {
            if (!results.find(function (r) {
              return r == recordValue;
            })) {
              results.push(recordValue);
            }
          }
        };

        for (_iterator12.s(); !(_step12 = _iterator12.n()).done;) {
          _loop2();
        }
      } catch (err) {
        _iterator12.e(err);
      } finally {
        _iterator12.f();
      }

      return results;
    }
  }, {
    key: "getNames",
    value: function getNames(recordArray) {
      var fieldName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'Name';
      var results = [];

      var _iterator14 = _createForOfIteratorHelper(recordArray),
          _step14;

      try {
        for (_iterator14.s(); !(_step14 = _iterator14.n()).done;) {
          var record = _step14.value;
          if (record) results.push(record.fields[fieldName]);
        }
      } catch (err) {
        _iterator14.e(err);
      } finally {
        _iterator14.f();
      }

      return results;
    }
  }, {
    key: "getFields",
    value: function getFields(recordArray) {
      var fieldName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'Name';
      var results = [];

      var _iterator15 = _createForOfIteratorHelper(recordArray),
          _step15;

      try {
        for (_iterator15.s(); !(_step15 = _iterator15.n()).done;) {
          var record = _step15.value;
          if (record.fields && record.fields[fieldName]) results.push(record.fields[fieldName]);
        }
      } catch (err) {
        _iterator15.e(err);
      } finally {
        _iterator15.f();
      }

      return this.deduplicate(results);
    }
  }, {
    key: "configure",
    value: function configure(_ref6) {
      var airKey = _ref6.airKey,
          airBase = _ref6.airBase,
          tables = _ref6.tables,
          tableQuery = _ref6.tableQuery,
          options = _ref6.options;
      Cytosis.airKey = airKey;
      Cytosis.airBase = {
        id: airBaseId
      };
      Cytosis.airBase.tables = tables || [];
      Cytosis.airBase.tableQuery = tableQuery || 'tables';
      Cytosis.airBase.options = options || 'Grid view';
    }
  }, {
    key: "split",
    value: function split(record, key) {
      var maxChunks = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 4;
      var chunkSize = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 100000;
      var itemString = JSON.stringify(record.fields[key]);
      if (itemString === undefined) return record;
      var parts = [];
      var i = 0;
      var length = itemString.length;

      while (length > 0) {
        parts.push(itemString.substr(i * chunkSize, chunkSize));
        length -= chunkSize;
        i++;
      }

      record.fields[key] = JSON.stringify({
        chunks: i,
        chunkSize: chunkSize
      });
      var j = 0;

      if (i < maxChunks) {
        while (j < i) {
          record.fields["".concat(key, "-").concat(j + 1)] = parts[j];
          j++;
        }
      } else {
        throw new Error("[Cytosis] \u2014 couldn\u2019t split record \"".concat(record.fields.Name, "\" \u2014 not enough chunks"));
      }

      return record;
    }
  }, {
    key: "unsplit",
    value: function unsplit(record, key) {
      if (!record.fields[key] || !JSON.parse(record.fields[key]).chunks) return JSON.parse(record.fields[key]);
      var chunks = JSON.parse(record.fields[key]).chunks;
      var itemString = '';
      var i = 0;

      while (i < chunks) {
        itemString += record.fields["".concat(key, "-").concat(i + 1)];
        delete record.fields["".concat(key, "-").concat(i + 1)];
        i++;
      }

      var data = JSON.parse(itemString);
      record.fields[key] = data;
      return record;
    }
  }, {
    key: "deduplicate",
    value: function deduplicate(recordArray) {
      return recordArray.filter(function (val, i, arr) {
        return arr.indexOf(val) == i;
      });
    }
  }, {
    key: "sort",
    value: function sort(recordArray) {
      var sortBy = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'Name';
      var sortFn = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : undefined;
      recordArray.sort(sortFn || function (a, b) {
        var nameA = a.fields[sortBy].toUpperCase();
        var nameB = b.fields[sortBy].toUpperCase();

        if (nameA < nameB) {
          return -1;
        }

        if (nameA > nameB) {
          return 1;
        }

        return 0;
      });
      return recordArray;
    }
  }, {
    key: "search",
    value: function search(str, sourceArray, searchByArray) {
      var linkedTableArray = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];
      if (!str) return sourceArray;
      return sourceArray.filter(function (obj) {
        var searchterm = str.toLowerCase();

        var _iterator16 = _createForOfIteratorHelper(searchByArray),
            _step16;

        try {
          for (_iterator16.s(); !(_step16 = _iterator16.n()).done;) {
            var field = _step16.value;

            if (obj.fields[field]) {
              if (typeof obj.fields[field] == 'string') {
                if (obj.fields[field].toLowerCase().includes(searchterm)) return true;
              } else if (Array.isArray(obj.fields[field])) {
                var _iterator17 = _createForOfIteratorHelper(obj.fields[field]),
                    _step17;

                try {
                  for (_iterator17.s(); !(_step17 = _iterator17.n()).done;) {
                    var strField = _step17.value;
                    if (strField.toLowerCase().includes(searchterm)) return true;
                  }
                } catch (err) {
                  _iterator17.e(err);
                } finally {
                  _iterator17.f();
                }

                if (linkedTableArray.length > 0) {
                  var _iterator18 = _createForOfIteratorHelper(linkedTableArray),
                      _step18;

                  try {
                    for (_iterator18.s(); !(_step18 = _iterator18.n()).done;) {
                      var linkedTable = _step18.value;
                      var records = Cytosis.getLinkedRecords(obj.fields[field], linkedTable, true);

                      var _iterator19 = _createForOfIteratorHelper(records),
                          _step19;

                      try {
                        for (_iterator19.s(); !(_step19 = _iterator19.n()).done;) {
                          var record = _step19.value;
                          if (record.fields['Name'].toLowerCase().includes(searchterm)) return true;
                        }
                      } catch (err) {
                        _iterator19.e(err);
                      } finally {
                        _iterator19.f();
                      }
                    }
                  } catch (err) {
                    _iterator18.e(err);
                  } finally {
                    _iterator18.f();
                  }
                }
              }
            }
          }
        } catch (err) {
          _iterator16.e(err);
        } finally {
          _iterator16.f();
        }

        return false;
      });
    }
  }, {
    key: "filter_or",
    value: function filter_or(keywords, field) {
      var orArr = [],
          strArr = "";
      keywords.map(function (keyword) {
        orArr.push("{".concat(field, "}=\"").concat(keyword, "\""));
      });
      strArr = orArr.join(', ');
      return "IF(OR(".concat(orArr, "), TRUE())");
    }
  }]);
  return Cytosis;
}();

var _default = Cytosis;
exports.default = _default;
